src/sentry/debug_files/artifact_bundles.py

```diff
@@ -139,3 +159,220 @@ def _index_urls_in_bundle(
+    was resolved with.
+    """
+
+    if debug_id:
+        bundles = get_artifact_bundles_containing_debug_id(project, debug_id)
+        if bundles:
+            return _maybe_renew_and_return_bundles(
+                {id: (date_added, "debug-id") for id, date_added in bundles}
+            )
+
+    total_release_bundles, not_indexed = get_release_indexing_state(project, release, dist)
+
+    if not total_release_bundles:
+        return []
+
+    # If all the bundles for this release are full indexed, we will only query
+    # the url index.
+    # Otherwise, if we are below the threshold, or only partially indexed, we
+    # want to return the N most recent bundles associated with the release,
+    # under the assumption that one of those should ideally contain the file we
+    # are looking for.
+    partially_indexed = total_release_bundles <= INDEXING_THRESHOLD
+
+    if total_release_bundles > INDEXING_THRESHOLD and not_indexed:
+        partially_indexed = True
+        metrics.incr("artifact_bundle_indexing.query_partial_index")
+        # TODO: spawn an async task to backfill non-indexed bundles
+        # lets do this in a different PR though :-)
+
+    # We keep track of all the discovered artifact bundles, by the various means of lookup.
+    # We are intentionally overwriting the `resolved` flag, as we want to rank these from
+    # coarse-grained to fine-grained.
+    artifact_bundles: Dict[int, Tuple[datetime, str]] = dict()
+
+    def update_bundles(bundles: Set[Tuple[int, datetime]], resolved: str):
+        for (bundle_id, date_added) in bundles:
+            artifact_bundles[bundle_id] = (date_added, resolved)
+
+    # First, get the N most recently uploaded bundles for the release,
+    # but only if the index is only partial:
+    if partially_indexed:
+        bundles = get_artifact_bundles_by_release(project, release, dist)
+        update_bundles(bundles, "release")
+
+    # Then, we are matching by `url`:
+    if url:
+        bundles = get_artifact_bundles_containing_url(project, release, dist, url)
+        update_bundles(bundles, "index")
+
+    return _maybe_renew_and_return_bundles(artifact_bundles)
+
+
+def get_release_indexing_state(project: Project, release_name: str, dist_name: str):
+    """
+    Returns the number of total bundles, and the number of not-yet-indexed bundles
+    associated with the given `release` / `dist`.
+    """
+    not_indexed = 0
+    total_bundles = 0
+
+    for state, count in (
+        ArtifactBundle.objects.filter(
+            organization_id=project.organization.id,
+            projectartifactbundle__project_id=project.id,
+            releaseartifactbundle__release_name=release_name,
+            releaseartifactbundle__dist_name=dist_name,
+        )
+        .values_list("indexing_state")
+        .annotate(count=Count("*"))
+    ):
+        if state == ArtifactBundleIndexingState.NOT_INDEXED.value:
+            not_indexed = count
+        total_bundles += count
```

Comment:
Do we want to work under the assumption than we will fade out the `NULL` values in the `indexing_state`?

Reply:
I switched the logic around, now counting the indexed bundles. Having `NULL` values all over the place is such a pain tbh.